### PR TITLE
Msdate nag

### PR DESCRIPTION
**Pull request by mevansgh1**
Issue #361 

nag msdate

### CODE REVIEW

File: docs-markdown/src/helper/utility.ts [hunk 1]
```diff
@@ -525,7 +527,7 @@ export function findReplacement(
 	if (result !== null && result.length) {
 		const groups = !isRegExp(expression) ? expression.groups : null;
 		const match = groups && result.groups ? result.groups[groups[0]] : result[0];
-		if (match) {
+		if (match && match !== value) {
```
Comment (mevansgh1) [inline docs-markdown/src/helper/utility.ts:530]: This is unrelated to my changes, just something I noticed.  There's no need to return a replacement if it's already the value we want to set.  This saves some branches of code from executing when someone attempts to update the ms.date when it's already the current date.  Just a nit pick fix

File: docs-markdown/src/helper/utility.ts [hunk 2]
```diff
@@ -442,7 +442,9 @@ export async function findAndReplaceTargetExpressions(event: TextDocumentChangeE
 					}
 				);
 
-				await applyReplacements(replacements, editor);
+				if (replacements.length > 0) {
```
Comment (mevansgh1) [inline docs-markdown/src/helper/utility.ts:445]: No need to call apply if there is nothing to apply

File: docs-markdown/src/test/suite/controllers/quick-pick-menu-controller.test.ts [hunk 3]
```diff
@@ -73,6 +73,14 @@ const context: ExtensionContext = {
 suite('Quick Pick Menu Controller', () => {
 	suiteSetup(() => {
 		sinon.stub(telemetry, 'sendTelemetryData');
+		sinon.stub(workspace, 'getConfiguration').returns({
```
Comment (mevansgh1) [inline docs-markdown/src/test/suite/controllers/quick-pick-menu-controller.test.ts:76]: There is a convention throughout the unit tests to override methods such as `getConfiguration()`.  However, this can result in unexpected behavior if someone is writing a new test and expecting the default behavior.  The use case I had was writing a new test case and expecting the configurations to be returned.  These overrides prevented my test from receiving the valid configuration.

For example, I have added config property 

```
{
  "markdown.metadataNag": true
}
```
to path `docs-markdown/src/test/data/repo/.vscode/settings.json`.  When I tried to load this using `workspace.getConfiguration()` it returns undefined
Screen Shot 2020-06-29 at 5 41 52 PM

By converting these overrides to stub we get the default behavior in new tests where it is defined.   You can see below how the property returns. This also allows us to not repeat ourselves with code in the unit tests.

Screen Shot 2020-06-29 at 5 51 58 PM

File: docs-markdown/src/test/suite/controllers/image-controller.test.ts [hunk 4]
```diff
@@ -44,17 +44,7 @@ const context: ExtensionContext = {
 	storagePath: '',
 	globalStoragePath: '',
 	logPath: '',
-	extensionUri: Uri.parse(uri),
-	environmentVariableCollection: {
```
Comment (mevansgh1) [inline docs-markdown/src/test/suite/controllers/image-controller.test.ts:48]: Fixes

Screen Shot 2020-06-29 at 6 37 00 PM

**Comment by mevansgh1** [comment]
I'm not sure why the build is failing it passes for me locally on both a Mac Os and Windows 10 VM.  Can someone please share the details of the build failure?

File: docs-markdown/src/helper/metadata.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import { findReplacement } from './utility';
+
+const blacklist: string[] = [];
+
+export async function nag() {
+	if (workspace.getConfiguration('markdown').metadataNag === false) {
+		return;
+	}
+	const editor = window.activeTextEditor;
+	if (!editor) {
```
Comment (nschonni) [inline docs-markdown/src/helper/metadata.ts:13]: Is this catching the case when doing a large Find/Replace?

Comment (mevansgh1) [reply docs-markdown/src/helper/metadata.ts:13]: Sorry I'm not sure I understand your question.   Can you please clarify? The line of code you commented on simply returns if there is no active editor.

This function is triggered from an onsave event so if someone were to do a large find/replace it would only trigger the function once IF auto-saved were enabled, assuming the find/replace is an atomic transaction.  

If auto save is not enabled then find/replace should not matter since this is only triggered when save is called.

**Comment by lamebrain** [comment]
@mevansgh1 Here's the error from the build:

Unable to write to Workspace Settings because markdown.metadataNag is not a registered configuration.

Error: Unable to write to Workspace Settings because markdown.metadataNag is not a registered configuration.
    at e.reject (file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:5363:839)
    at e.resolveAndValidate (file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:5366:717)
    at e.doWriteConfiguration (file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:5361:70)
    at Object.factory (file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:5360:801)
    at u.consume (file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:130:75)
    at file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:129:886
    at new Promise (<anonymous>)
    at u.queue (file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:129:810)
    at e.writeConfiguration (file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:5360:786)
    at C.writeConfigurationValue (file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:5389:83)
    at file:///D:/a/1/s/docs-markdown/.vscode-test/vscode-1.46.1/resources/app/out/vs/workbench/workbench.desktop.main.js:5377:320
    at processTicksAndRejections (internal/process/task_queues.js:85:5)

File: docs-markdown/src/helper/metadata.ts (hunk 5 above)
Comment (nschonni) [reply docs-markdown/src/helper/metadata.ts:13]: I was wondering if this section of code was protecting against the nag running on a large Find/Replace when run in VS Code. As a fake example, if you do a find for the letter `a`, and replace it with `b`, are you going to have 100+ nags stacked on each other?

File: docs-markdown/src/test/suite/controllers/image-controller.test.ts (hunk 4 above)
Comment (mevansgh1) [reply docs-markdown/src/test/suite/controllers/image-controller.test.ts]: This error actually was related to me having a newer version of the `@types/vscode` package because my package-lock.json had a different version in it then what was in package.json.  

In this repository I see 8 extensions, 6 of which commit their package-lock.json.  Only the *docs-images* and *docs-markdown* extensions do **not** commit their package-lock.json.

per the documentation [configuring-npm](https://docs.npmjs.com/configuring-npm/package-lock-json.html)

`This file is intended to be committed into source repositories, and serves various purposes`

I have added a new commit to this PR to commit this file.  I'm not sure why this was originally .gitignored, but please review this comment.

File: docs-markdown/src/helper/metadata.ts (hunk 5 above)
Comment (mevansgh1) [reply docs-markdown/src/helper/metadata.ts:13]: I tested this locally with auto-save enabled and vscode only triggers 1 onSave event when running a find/replace on the same string. I did add code below to make sure we only nag once per file in a session.  So this code will also prevent the issue you're describing IF vscode did trigger 100 onsave events during a find/replace.

```
// skip files that have already been saved
	if (blacklist.indexOf(fileName) > -1 === true) {
		return;
	}
	blacklist.push(fileName);
```

File: docs-markdown/src/test/suite/controllers/image-controller.test.ts (hunk 4 above)
Comment (mevansgh1) [reply docs-markdown/src/test/suite/controllers/image-controller.test.ts]: disregard will submit in separate PR

**Comment by jamarw** [comment]
Hi @mevansgh1 , just wanted to let you know we're reviewing this PR now.  Thanks!

File: docs-markdown/src/helper/metadata.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+		return;
+	}
+
+	if (!isMarkdownFileCheck(editor, false)) {
+		return;
+	}
+
+	const fileName = editor.document.fileName;
+	// skip files that have already been saved
+	if (blacklist.indexOf(fileName) > -1 === true) {
```
Comment (IEvangelist) [inline docs-markdown/src/helper/metadata.ts:24]: Please reconsider naming this variable to something other than `blacklist`, perhaps `disallowlist`?

File: docs-markdown/src/helper/metadata.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+		const msDateRegex = /ms.date:\s*\b(.+?)$/im;
+		const msDateReplacement = findReplacement(
+			editor.document,
+			content,
+			`ms.date: ${toShortDate(new Date())}`,
+			msDateRegex
+		);
+		if (msDateReplacement) {
+			const syncDate = await window.showInformationMessage(
+				"Would you like to update ms.date to today's date?",
```
Comment (IEvangelist) [inline docs-markdown/src/helper/metadata.ts:40]:
```suggestion
				"Would you like to update `ms.date` to today's date?",
```

I believe that we can format this as code, and it will render as such in the prompt.

File: docs-markdown/src/helper/metadata.ts (hunk 6 above)
Comment (jamarw) [reply docs-markdown/src/helper/metadata.ts:24]: Hi @IEvangelist , good call. I believe there's a feature/work item to rename all of these instances soon.  Thanks
